specifyweb/frontend/js_src/lib/components/SchemaConfig/UniquenessRules.tsx

```diff
@@ -0,0 +1,419 @@
+            <td>{schemaText.scope()}</td>
+          </tr>
+        </thead>
+        {tableRules?.map(({ rule, duplicates }, index) => (
+          <UniquenessRuleRow
+            container={container}
+            fetchedDuplicates={duplicates}
+            fields={allFieldNames}
+            key={index}
+            label={getUniqueInvalidReason(
+              rule.scopes.map(
+                (scope) =>
+                  (getTable(container.name)?.getField(scope.name) ??
+                    '') as Relationship
+              ),
+              filterArray(
+                rule.fields.map((field) =>
+                  getTable(container.name)?.getField(field.name)
+                )
+              )
+            )}
+            relationships={relationships}
+            rule={rule}
+            onChange={(newRule): void => handleRuleValidation(newRule, index)}
+            onRemoved={(): void => setTableRules(removeItem(tableRules, index))}
+          />
+        ))}
+      </table>
+      <Button.Small
+        className="w-fit"
+        onClick={(): void =>
+          handleRuleValidation(
+            {
+              id: null,
+              fields: [fields[0]],
+              isDatabaseConstraint: false,
+              scopes: [],
+            },
+            tableRules.length
+          )
+        }
+      >
+        {schemaText.addUniquenessRule()}
+      </Button.Small>
+    </Dialog>
+  );
+}
+
+function UniquenessRuleRow({
+  rule,
+  container,
+  label,
+  fields,
+  relationships,
+  fetchedDuplicates,
+  onChange: handleChanged,
+  onRemoved: handleRemoved,
+}: {
+  readonly rule: UniquenessRule;
+  readonly container: SerializedResource<SpLocaleContainer>;
+  readonly label: string;
+  readonly fields: RA<
+    SerializedResource<SpLocaleContainerItem> & WithFetchedStrings
+  >;
+  readonly relationships: RA<
+    SerializedResource<SpLocaleContainerItem> & WithFetchedStrings
+  >;
+  readonly fetchedDuplicates: UniquenessRuleValidation | undefined;
+  readonly onChange: (newRule: typeof rule) => void;
+  readonly onRemoved: () => void;
+}): JSX.Element {
+  const readOnly = rule.isDatabaseConstraint;
+  const [separator] = userPreferences.use(
+    'queryBuilder',
+    'behavior',
+    'exportFileDelimiter'
+  );
+
+  const uniqueFields = React.useMemo(
+    () => fields.map((field) => [field.name, field.strings.name.text] as const),
+    [fields]
+  );
+
+  const [advancedScopes, hierarchyScopes] = React.useMemo(
+    () =>
+      split(relationships, (field) =>
+        field.name === databaseFieldName
+          ? true
+          : schema.orgHierarchy.includes(
+              (strictGetTable(container.name)
+                .getRelationship(field.name)
+                ?.getReverse()?.table.name ??
+                '') as typeof schema.orgHierarchy[number]
+            )
+      ).map((scopes) =>
+        scopes.map((field) => [field.name, field.strings.name.text] as const)
+      ),
+    [container.name, relationships]
```

Comment:
`[relationships]` is enough, so also get rid of `container.name` prop. Notice relationships get updated in parent when container.name changes.

Generally, aim for components that have as little context as possible so that they can be used in other places, and try to reason through if dependencies are really needed, and if some other hook updates it or not (which happens in our case)

Reply:
Fixed in https://github.com/specify/specify7/pull/4215/commits/0e4a7f874cdbf98d06e15b5845ef64904410fc94